internal/runbits/checkout/checkout.go

```diff
@@ -75,102 +74,115 @@ func (r *Checkout) Run(ns *project.Namespaced, branchName, cachePath, targetPath
 	proj := ns.Project
 	commitID := ns.CommitID
 	var language string
-	if !fromArchive {
-		// If project does not exist at path then we must checkout
-		// the project and create the project file
-		pj, err := model.FetchProjectByName(ns.Owner, ns.Project, r.auth)
+	if !onlyValidate {
+		owner, proj, commitID, branchName, language, err = r.checkout(path, ns, branchName, commitID, cachePath, noClone)
 		if err != nil {
-			return "", locale.WrapError(err, "err_fetch_project", "", ns.String())
+			return "", errs.Wrap(err, "Unable to checkout project")
 		}
-		proj = pj.Name
-
-		var branch *mono_models.Branch
-
-		switch {
-		// Fetch the branch the given commitID is on.
-		case commitID != nil:
-			for _, b := range pj.Branches {
-				if belongs, err := model.CommitBelongsToBranch(ns.Owner, ns.Project, b.Label, *commitID, r.auth); err == nil && belongs {
-					branch = b
-					break
-				} else if err != nil {
-					return "", errs.Wrap(err, "Could not determine which branch the given commitID belongs to")
-				}
-			}
-			if branch == nil {
-				return "", &errCommitDoesNotBelong{CommitID: *commitID}
-			}
+	} else if commitID == nil {
+		return "", errNoCommitID
+	}
 
-		// Fetch the given project branch.
-		case branchName != "":
-			branch, err = model.BranchForProjectByName(pj, branchName)
-			if err != nil {
-				return "", locale.WrapError(err, "err_fetch_branch", "", branchName)
-			}
-			commitID = branch.CommitID
+	if err := CreateProjectFiles(path, cachePath, owner, proj, branchName, commitID.String(), language); err != nil {
+		return "", errs.Wrap(err, "Could not create project files")
+	}
 
-		// Fetch the default branch for the given project.
-		default:
-			branch, err = model.DefaultBranchForProject(pj)
-			if err != nil {
-				return "", errs.Wrap(err, "Could not grab branch for project")
-			}
-			commitID = branch.CommitID
+	if r.config.GetBool(constants.OptinBuildscriptsConfig) {
+		if err := buildscript_runbit.Initialize(path, r.auth); err != nil {
+			return "", errs.Wrap(err, "Unable to initialize buildscript")
 		}
-		commitID = branch.CommitID
-		branchName = branch.Label
+	}
 
-		if commitID == nil {
-			return "", errNoCommitID
-		}
+	return path, nil
+}
 
-		// Clone the related repo, if it is defined
-		if !noClone && pj.RepoURL != nil && *pj.RepoURL != "" {
-			err := r.repo.CloneProject(ns.Owner, ns.Project, path, r.Outputer, r.analytics)
-			if err != nil {
-				return "", locale.WrapError(err, "err_clone_project", "Could not clone associated git repository")
+// checkout performs the checkout and returns the checkout's owner, project, commitID, branchName,
+// and language.
+func (r *Checkout) checkout(
```

Comment:
It doesn't feel right to call this `checkout`. That's what the main `Run` function does (ie. creating the project file is effectively the checkout).

Suggest turning this into a function called `sanitize()` that does all the cleaning up of parameters, and then separately call the clone behaviour from `Run()`.

I'm not a big fan of the number of parameters this returns, but I also don't see an obvious solution that doesn't add boilerplate that I'm even less of a fan of. Just flagging it in case you have ideas.

Reply:
I've gone with `fetchProject` since I feel like that is the goal of this function. Fetch the project from the platform and return its complete set of parameters.